zetaclient/evm/evm_signer.go

```diff
@@ -324,246 +332,233 @@ func (signer *Signer) SignCommandTx(
-		}
-	}
+func SetupGas(
+	cctx *types.CrossChainTx,
+	logger zerolog.Logger,
+	client interfaces.EVMRPCClient,
+	chain *common.Chain,
+	txData *TransactionData) error {
 
-	gasLimit := cctx.GetCurrentOutTxParam().OutboundTxGasLimit
-	if gasLimit < 100_000 {
-		gasLimit = 100_000
-		logger.Warn().Msgf("gasLimit %d is too low; set to %d", cctx.GetCurrentOutTxParam().OutboundTxGasLimit, gasLimit)
+	txData.gasLimit = cctx.GetCurrentOutTxParam().OutboundTxGasLimit
+	if txData.gasLimit < 100_000 {
+		txData.gasLimit = 100_000
+		logger.Warn().Msgf("gasLimit %d is too low; set to %d", cctx.GetCurrentOutTxParam().OutboundTxGasLimit, txData.gasLimit)
 	}
-	if gasLimit > 1_000_000 {
-		gasLimit = 1_000_000
-		logger.Warn().Msgf("gasLimit %d is too high; set to %d", cctx.GetCurrentOutTxParam().OutboundTxGasLimit, gasLimit)
+	if txData.gasLimit > 1_000_000 {
+		txData.gasLimit = 1_000_000
+		logger.Warn().Msgf("gasLimit %d is too high; set to %d", cctx.GetCurrentOutTxParam().OutboundTxGasLimit, txData.gasLimit)
 	}
 
-	logger.Info().Msgf("chain %s minting %d to %s, nonce %d, finalized zeta bn %d", toChain, cctx.InboundTxParams.Amount, to.Hex(), nonce, cctx.InboundTxParams.InboundTxFinalizedZetaHeight)
-	sendHash, err := hex.DecodeString(cctx.Index[2:]) // remove the leading 0x
-	if err != nil || len(sendHash) != 32 {
-		logger.Error().Err(err).Msgf("decode CCTX %s error", cctx.Index)
-		return
-	}
-	var sendhash [32]byte
-	copy(sendhash[:32], sendHash[:32])
-
-	// use dynamic gas price for ethereum chains
-	var gasprice *big.Int
-
+	// use dynamic gas price for ethereum chains.
 	// The code below is a fix for https://github.com/zeta-chain/node/issues/1085
 	// doesn't close directly the issue because we should determine if we want to keep using SuggestGasPrice if no OutboundTxGasPrice
 	// we should possibly remove it completely and return an error if no OutboundTxGasPrice is provided because it means no fee is processed on ZetaChain
 	specified, ok := new(big.Int).SetString(cctx.GetCurrentOutTxParam().OutboundTxGasPrice, 10)
 	if !ok {
-		if common.IsEthereumChain(toChain.ChainId) {
-			suggested, err := signer.client.SuggestGasPrice(context.Background())
+		if common.IsEthereumChain(chain.ChainId) {
+			suggested, err := client.SuggestGasPrice(context.Background())
 			if err != nil {
-				logger.Error().Err(err).Msgf("cannot get gas price from chain %s ", toChain)
-				return
+				return errors.Join(err, fmt.Errorf("cannot get gas price from chain %s ", chain))
 			}
-			gasprice = roundUpToNearestGwei(suggested)
+			txData.gasPrice = roundUpToNearestGwei(suggested)
 		} else {
-			logger.Error().Err(err).Msgf("cannot convert gas price  %s ", cctx.GetCurrentOutTxParam().OutboundTxGasPrice)
-			return
+			return fmt.Errorf("cannot convert gas price  %s ", cctx.GetCurrentOutTxParam().OutboundTxGasPrice)
 		}
 	} else {
-		gasprice = specified
+		txData.gasPrice = specified
 	}
+	return nil
+}
+
+func SetTransactionData(
```

Comment:
The function is doing some validation as well as setting up the data structure. NewTransationData would imply a simple constructor.

Reply:
You can have preprocessing or validation in a construction. Referring to the code it seems to be the actually entrypoint to initialize the structure:
```
	// Setup Transaction input
	txData := BaseTransactionData{}
	txData.height = height
	skipTx, err := txData.SetTransactionData(cctx, evmClient, signer.client, logger)
```
It would be better to make clear it's the entrypoint to initialize the structure with a new method that return the object.
This makes code maintainability easier and also mutable function are harder to debug